feat: automatic reload when file changes externally

AnthonyMichaelTDM opened the pull request:
Fixes #1125 

Implements the idea raised by @porridgewithraisins in this comment: https://github.com/helix-editor/helix/issues/1125#issuecomment-3073782827

The implementation was modeled after the implementation of the autosave feature.

## pros of this approach

- The code is fairly simple, only a few hundred lines, and is based off of the implementations of similar features. This all reduces the maintenance burden of accepting this PR
- Cross platform. Any platform that stores when a file was last modified in its metadata will be supported by this implementation.
- Uses the existing event system instead of bringing in another runtime.

## con of this approach

- using the event system to perform an action at a periodic interval was a little convoluted at first, but I don't think it ended up being that bad

## Configuration examples

Here are some example snippets of configuration for this feature:

- disable auto-reloading (default)
```toml
[editor]
auto-reload = false
```
or
```toml
[editor.auto-reload]
focus-gained = false
```
- auto-reload on focus
```toml
[editor]
auto-reload = true
```
or
```toml
[editor.auto-reload]
focus-gained = true
```
- auto-reload at some periodically at time interval (5 seconds in this example)
```toml
[editor.auto-reload]
periodic.enable = true
periodic.interval = 5000
```
- of course, you could have it reload on focus and at an interval too:
```toml
[editor.auto-reload]
focus-gained = true
periodic.enable = true
periodic.interval = 5000
```

## future work

- add a command to toggle this at runtime if, say, you're working with multiple very large files.

AnthonyMichaelTDM committed:
feat: automatic reload when file changes externally

Implements the idea raised by @porridgewithraisins in this comment:
https://github.com/helix-editor/helix/issues/1125#issuecomment-3073782827

The implementation was modeled after the implementation of the autosave feature.

Here are some example snippets of configuration for this feature:

- disable auto-reloading (default)
```toml
[editor]
auto-reload = false
```
or
```toml
[editor.auto-reload]
focus-gained = false
```
- auto-reload on focus
```toml
[editor]
auto-reload = true
```
or
```toml
[editor.auto-reload]
focus-gained = true
```
- auto-reload at some periodically at time interval (5 seconds in this example)
```toml
[editor.auto-reload]
periodic.enable = true
periodic.interval = 5000
```
- of course, you could have it reload on focus and at an interval too:
```toml
[editor.auto-reload]
focus-gained = true
periodic.enable = true
periodic.interval = 5000
```

AnthonyMichaelTDM committed:
fix: typos

I has originally named this feature `auto_read`, but changed the name to `auto_reload`, this changes some references I forgot to update to the new name

AnthonyMichaelTDM commented:
As far as changes to the existing API go: the `reload_all` command had its visibility changed to `pub`, the `Document` struct has a new function (`get_last_saved_time`) (the alternative was to change the visibility of the `last_saved_time` field), and `AutoReload` configuration was added to the editor configuration

AnthonyMichaelTDM commented on helix-term/src/ui/editor.rs:
```diff
@@ -1468,6 +1468,12 @@ impl Component for EditorView {
             Event::Mouse(event) => self.handle_mouse_event(event, &mut cx),
             Event::IdleTimeout => self.handle_idle_timeout(&mut cx),
             Event::FocusGained => {
+                if context.editor.config().auto_reload.focus_gained {
+                    helix_event::send_blocking(
+                        &context.editor.handlers.auto_reload,
+                        helix_view::handlers::AutoReloadEvent::EditorFocused,
+                    );
+                }
```
not working as expected, will fix

AnthonyMichaelTDM committed:
fix: auto-reloading on focus

Open question: should we prompt users here too? currently don't

AnthonyMichaelTDM committed:
doc(auto-reloading): document configuration of auto-reloading

Axlefublr commented:
this is wonderful :D I was sorta hoping for a file watching type of solution, rather than a time-based one, but this is still very helpful!

Axlefublr commented:
my guess is that while file watching is viable on linux, on windows and macos it's not something that's easily provided? or hell, it might even be a filesystem thing

AnthonyMichaelTDM commented:
This approach is arguably not as elegant, but it is much more simple, integrates with the existing event system, requires no additional dependencies, has negligible performance overhead (all it's really doing is checking file metadata every so often), and is just as cross-platform as [std::fs::metadata::modified](https://doc.rust-lang.org/std/fs/struct.Metadata.html#method.modified)

A few other people have attempted to add file-watching (notify) based implementations already, so I wanted to provide a different approach.
I have no preference as to which approach is eventually merged. I just want an automatic reload feature, and I feel like providing the helix maintainers with more options is the best path toward reaching that goal.

Axlefublr commented:
good thinking! even a more “naive” implementation like this will solve most of reloading needs as is. hope this gets attention!

Tudyx commented:
You don't have to close and reopen helix, you can use the `:reload-all` (or `:rla`) when you switch `git` branch

AnthonyMichaelTDM commented:
Is there someone I can ping to get a proper review? Feedback seems positive, and AFAIK the problem hasn't been fixed by a merged PR.

RoloEdits commented:
https://github.com/helix-editor/helix/pull/14544 was opened up by a maintainer that would appear to subsume this.

archseer commented:
Replaced by #14544
